src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+				currentRotation.casts.findIndex(cast => cast.ability.guid === ACTIONS.DESPAIR.id)) {
+				this.manafontBeforeDespair++
+				currentRotation.errorCode = ERROR_CODES.MANAFONT_BEFORE_DESPAIR
+			}
+
+		// Check if the rotation included more than one Fire 1
+		const fire1Count = currentRotation.casts.filter(cast => cast.ability.guid === ACTIONS.FIRE_I.id).length
+		if (fire1Count > 1) {
+			currentRotation.errorCode = ERROR_CODES.EXTRA_F1
+			this.extraF1s += Math.max(0, fire1Count-1)
+		}
+
+		// If this cycle ends with downtime or is the last cycle, many of the errors we normally check for
+		// don't apply, so split the processing pathway here
+		if (currentRotation.finalOrDowntime) {
+			this.processDowntimeCycle(currentRotation)
+			return
+		}
+
+		this.processNormalCycle(currentRotation)
+	}
+
+	// Process errors for a normal mid-fight cycle
+	private processNormalCycle(currentRotation: Cycle) {
+		// Check to make sure we didn't lose Fire 4 casts due to spending MP on T3 hardcasts
+		const hardT3Count = currentRotation.casts.filter(cast => cast.ability.overrideAction)
+			.filter(cast => cast.ability.overrideAction.id === ACTIONS.THUNDER_III_FALSE.id).length
+		if (hardT3Count > 1 || (hardT3Count > 0 && currentRotation.firePhaseStartMP < MIN_MP_FOR_FULL_ROTATION)) {
+			this.extraT3s++
+			currentRotation.errorCode = ERROR_CODES.EXTRA_T3
+		}
+
+		// Why so icemage?
+		if (!currentRotation.casts.filter(cast => FIRE_SPELLS.includes(cast.ability.guid)).length) {
+			this.rotationsWithoutFire++
+			currentRotation.errorCode = ERROR_CODES.NO_FIRE_SPELLS
+		}
+
+		// If they're just missing Fire 4 because derp, note it
+		if (currentRotation.missingFire4s) {
+			currentRotation.errorCode = ERROR_CODES.MISSING_FIRE4S
+		}
+	}
+
+	// Process errors for a cycle that was cut short by downtime or by the fight ending
+	private processDowntimeCycle(currentRotation: Cycle) {
+		currentRotation.errorCode = ERROR_CODES.FINAL_OR_DOWNTIME
+
+		// Check if more Fire 4s could've been cast by skipping Blizzard 4 before this downtime
+		if (currentRotation.gaugeStateBeforeFire.umbralHearts > 0 && currentRotation.missingFire4s === 2) {
+			this.missedF4s++
+		}
+
+		// TODO: Check for hardcast T3s, if this cycle ends in downtime, that cast time should've been a fire spell
+	}
+
+	output() {
+		const outliers: Cycle[] = this.history.filter(cycle => cycle.errorCode >
+			ERROR_CODES.SHORT || DEBUG_SHOW_ALL_CYCLES)
+		if (outliers.length > 0 ) {
+			return <Fragment>
+				<Message>
+					<Trans id="blm.rotation-watchdog.rotation-table.message">
+						The core of BLM consists of six <ActionLink {...ACTIONS.FIRE_IV} />s and one <ActionLink {...ACTIONS.DESPAIR} /> per rotation (seven <ActionLink {...ACTIONS.FIRE_IV} />s and two <ActionLink {...ACTIONS.DESPAIR} />s with <ActionLink {...ACTIONS.MANAFONT} />).<br/>
+						Avoid missing Fire IV casts where possible.
+					</Trans>
+				</Message>
+			<Message warning icon>
```

Comment:
not super keen on two messages after one another but i guess it's 'aite. this one's indentation is jank.

i _am_ trying to slowly remove semantic, so if this can be done with the in-house message that'd be 👌

Reply:
switching to the /akkd/Message component removed the outlines/separation it currently has and it looked like hot garbage. Can plop some screenshots in Discord to show you what I mean if you want. I did fix the wonky indentation though.